### ISSUE TITLE
Code blocks in MarkupContent are always highlighted as Markdown code

### ISSUE THREAD

**Issue by Kha** [emacs-lsp/lsp-mode#2588]
When rendering markdown inside MarkupContent, `markdown-fontify-code-block-default-mode ` is always set to `gfm-view-mode` (or `gfm-mode`) since that is the major mode activated just before calling `lsp--setup-markdown`:
https://github.com/emacs-lsp/lsp-mode/blob/62cd1b2e569c72638ba4bd42a0290192c224c28d/lsp-mode.el#L4723-L4727

However, saving `major-mode` before switching to the major mode is not sufficient since this whole function is called inside a temp buffer in `lsp--fontlock-with-mode`:
https://github.com/emacs-lsp/lsp-mode/blob/62cd1b2e569c72638ba4bd42a0290192c224c28d/lsp-mode.el#L4789-L4791
So at the beginning of `lsp--setup-markdown`, `major-mode` is already set to `fundamental-mode` instead of the desired mode of the actual buffer. Here `mode` is `lsp--setup-markdown`, which was retrieved from `lsp-language-id-configuration` when passing the MarkupContent's "kind" `"markdown"` to `lsp--render-string`:
https://github.com/emacs-lsp/lsp-mode/blob/62cd1b2e569c72638ba4bd42a0290192c224c28d/lsp-mode.el#L4816-L4817

**Comment by yyoncho** [comment]
I am not sure I understand the issue. What is the current behaviour and what is the expected behaviour in this case?

**Comment by Kha** [comment]
Yeah, let me try again with a more high-level description: given the response
````
[Trace - 10:55:32 PM] Received response 'textDocument/hover - (223)' in 2ms.
Result: {
  "contents": {
    "kind": "markdown",
    "value": "```lean\nEq.trans : ∀ {α : Type} {a b c : α}, a = b → b = c → a = c\n```"
  },
  "range": {
    "end": {
      "character": 16,
      "line": 17
    },
    "start": {
      "character": 8,
      "line": 17
    }
  }
}
````
what I get in both the lsp-ui hover and eldoc is
![image](https://user-images.githubusercontent.com/109126/107160590-7f33a000-6997-11eb-85dc-604b0bb4b0de.png)
while what I would have expected is
![image](https://user-images.githubusercontent.com/109126/107160921-7a6feb80-6999-11eb-820b-3823cab883ff.png)
which I get by "fixing" the code by hard-coding my major mode, and also explicitly adding its language ID to `markdown-code-lang-modes`:
```diff
modified   lsp-mode.el
@@ -4667,7 +4667,9 @@ point."
 (defun lsp--setup-markdown (mode)
   "Setup the ‘markdown-mode’ in the frame.
 MODE is the mode used in the parent frame."
+  (setq mode 'lean4-mode)
   (make-local-variable 'markdown-code-lang-modes)
+  (add-to-list 'markdown-code-lang-modes ("lean" . lean4-mode)) ; same pair as in `lsp-language-id-configuration`
```
I don't know if it's more correct/appropriate for the language server to return code blocks with or without language specifier, but in either case there seems to be an issue:
* code blocks without specified language are highlighted using the `gfm-view-mode` syntax table as described in the original issue
* code blocks with specified language are not highlighted at all unless both the issue above is fixed and the language ID is added to `lsp-custom-markup-modes`. The latter step should probably be automatic.

I assume that the first issue is unintended since `lsp--setup-markdown` explicitly says "MODE is the mode used in the parent frame.", which is not the case right now.

**Comment by yyoncho** [comment]
Got it, there is easy dirty fix, I will see if we can do better.

**Comment by yyoncho** [comment]
proposed fix at #2666

**Comment by Kha** [comment]
Nice, it works now after registering the language ID as a markdown language:
```
(add-to-list 'lsp-custom-markup-modes '(lean4-mode "lean")) ; same pair as in `lsp-language-id-configuration`
```
What do you think, should this be done automatically?

**Comment by yyoncho** [comment]
> What do you think, should this be done automatically?

IMHO better to do a PR against `markdown-mode` to add that pair, thus it will work fine even out of `lsp-mode`.
